src/main/java/com/guciowons/shoppingbasket/Basket/BasketService.java

```diff
@@ -1,67 +1,71 @@
+        basketRepository.save(newBasket);
         return newBasket;
     }
 
+
     public BasketSummarized addProductToBasket(int basketId, int productId, int quantity) throws NoExternalConnectionException, NoProductException, NoBasketException{
-        try{
-            return basketDao.findById(basketId)
-                    .map(basket -> productService.getProductById(productId)
-                            .map(product -> {
-                                basket.addProduct(productId, quantity);
-                                System.out.println(product);
-                                return basketSummarizer.summarizeBasket(basket.getContent(), productService.getProducts());
-                            })
-                            .orElseThrow(() -> new NoProductException("No such product!")))
-                    .orElseThrow(() -> new NoBasketException("No such basket"));
-        }catch (FeignException e){
-            throw new NoExternalConnectionException("Cant connect with external api");
-        }
+        return basketRepository.findById(basketId)
+                .map(basket -> addProductIfExists(productService.getProductById(productId), basket, quantity))
+                .orElseThrow(() -> new NoBasketException("No such basket"));
     }
 
+    private BasketSummarized addProductIfExists(Optional<Product> optionalProduct, Basket basket, int quantity){
+        return optionalProduct
+                .map(product -> {
+                    basket.addProduct(product.getId(), quantity);
+                    basketRepository.save(basket);
+                    return basketSummarizer.summarizeBasket(basket.getContent(), productService.getProducts());
+                })
+                .orElseThrow(() -> new NoProductException("No such product!"));
+    }
+
+
     public void removeProductFromBasket(int basketId, int productId, int quantity) throws NoExternalConnectionException, NoProductInBasketException, NoProductException, NoBasketException {
-        try {
-            basketDao.findById(basketId).ifPresentOrElse(
-                    basket -> productService.getProductById(productId).ifPresentOrElse(
-                            product -> basket.removeProduct(productId, quantity),
-                            () -> {
-                                throw new NoProductException("No such product!");
-                            }),
+        basketRepository.findById(basketId).ifPresentOrElse(
+                basket -> removeProductIfExists(productService.getProductById(productId), basket, quantity),
                     () -> {
                         throw new NoBasketException("No such basket!");
                     });
-        }catch (FeignException e){
-            throw new NoExternalConnectionException("Cant connect with external api");
-        }
     }
 
+    private void removeProductIfExists(Optional<Product> optionalProduct, Basket basket, int quantity){
```

Comment:
Przekazywanie Optionala jako parametr metody jest antypatternem, poczytaj sobie dlaczego, nawet IntelliJ oznacza to na żółto jako potencjalne problemy w klasie.

Powinieneś przekazać po prostu `productId` i całą logikę pobierania zawrzeć w tej metodzie. Albo wcześniej sprawdzić czy ten obiekt istnieje i rzucić wyjątek jeśli nie.